optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -58,55 +45,283 @@ that uses a functional approach:
+                .join(Employee.class) // Returns BiStream<Shift, Employee>.
+                .groupBy((shift, employee) -> employee) // Returns UniStream<Employee>.
+                ...
+    }
+----
+
+The following constraint stream cardinalities are currently supported.
+
+|===
+|Cardinality|Prefix|Defining interface
+|1          |   Uni|`UniConstraintStream<A>`
+|2          |    Bi|`BiConstraintStream<A, B>`
+|3          |   Tri|`TriConstraintStream<A, B, C>`
+|4          |  Quad|`QuadConstraintStream<A, B, C, D>`
+|===
+
+
+[[constraintStreamsBuildingBlocks]]
+== Building Blocks
+
+Constraint streams are chains of different operations, also called building blocks.
+Each constraint stream starts with a `from(...)` building block and is terminated by either a penalty or a reward.
+This is an example of the simplest possible constraint stream:
+
+[source,java,options="nowrap"]
+----
+    private Constraint penalizeInitializedShifts(ConstraintFactory factory) {
+        return factory.from(Shift.class)
+                .penalize("Initialized shift", HardSoftScore.ONE_SOFT);
+    }
+----
+
+This constraint stream will go over all known and initialized instances of `Shift`.
+To include instances yet uninitialized, replace the `from()` building block with `fromUnfiltered()`:
+
+[source,java,options="nowrap"]
+----
+    private Constraint penalizeAllShifts(ConstraintFactory factory) {
+        return factory.fromUnfiltered(Shift.class)
+                .penalize("A shift", HardSoftScore.ONE_SOFT);
     }
 ----
 
-image::ConstraintStreams/constraintStreamIntroduction.png[align="center"]
-image::ConstraintStreams/constraintStreamIncrementalCalculation.png[align="center"]
-image::ConstraintStreams/constraintStreamJustification.png[align="center"]
+
+[[constraintStreamsPenaltiesRewards]]
+=== Penalties And Rewards
+
+The purpose of constraint streams is to build up a <<whatIsAScore,score>> for a <<planningProblemAndPlanningSolution,solution>>.
+To accomplish that, every constraint stream must be terminated by a call to either a `penalize()` or a `reward()`
+building block.
+As their names suggest, the former makes the score worse while the latter improves it.
+Penalties and rewards have several components:
+
+Constraint name::
+Human readable descriptive name for the constraint, which must be unique within the entire `ConstraintProvider`.
+Constraint weight::
+A constant score value indicating how much every breach of the constraint affects the score.
+Valid examples include `SimpleScore.ONE`, `HardSoftScore.ONE_HARD`, `HardMediumSoftScore.of(1, 2, 3)`, ...
+Constraint match weigher (Optional)::
+A function indicating how many times the constraint weight should be applied in the score.
+This number is typically determined from the contents of the constraint stream.
+Defaults to 1.
+Constraint package (Optional)::
+Java package that the constraint resides in.
+Defaults to the package where the `ConstraintProvider` resides.
```

Comment:
IIRC, it does unless there's a @ConstraintConfiguration class in which case it takes that package for DRL compatibility... Check the code on this.

PS: this could be one of those times that falls under "now that I have to document it I am not so sure any more if it was a good idea to code it like that"